### PR TITLE
refactor: apply webpack-defaults

### LINKED ISSUE #52: webpack-defaults upgrade
Addition of `webpack-defaults` & associated refactoring as a part of the next `Major` release

Issue exists for status tracking across the organization. 

>Please do not close

### PR DESCRIPTION
**Pull request by d3viant0ne**
Intended to be merged before & released with #26 & #45 as a part of `1.0.0` on a beta `dist-tag` once this has been finished and properly tested.

This is as close to inline with the defaults eslint config as this is going to get without making non-trivial structural changes to the lib which isn't something we should really be doing without a proper set of validations.

```

- refactor: Apply webpack-defaults 

BREAKING CHANGE: Enforces `peerDependencies` of `"webpack": ">= 3.0.0-rc.0 || ^3.0.0"`. 

BREAKING CHANGE: Enforces `engines` of `"node": ">=4.3.0 < 5.0.0 || >= 5.10`

BREAKING CHANGE: Remove loose dependency on Node Zopfli, which has been extracted to it's own plugin https://github.com/webpack-contrib/zopfli-webpack-plugin
```

Closes #52

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import async from 'async';
+import RawSource from 'webpack-sources/lib/RawSource';
+
+class CompressionPlugin {
+  constructor(options = {}) {
+    this.asset = options.asset || '[path].gz[query]';
+    this.algorithm = options.algorithm || 'gzip';
+    this.filename = options.filename || false;
+    this.compressionOptions = {};
+    if (typeof this.algorithm === 'string') {
```
Comment (bebraw) [inline src/index.js]: I would use more newlines before ifs but maybe that's just me. 👍

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          blocksplittinglast: hasOwnProperty.call(options, 'blocksplittinglast') ? options.blocksplittinglast : false,
+          blocksplittingmax: options.blocksplittingmax ? options.blocksplittingmax : 15,
+        };
+        this.algorithm = (content, options, fn) => {
+          zopfli.gzip(content, options, fn);
+        };
+      } else {
+        const zlib = require('zlib');
+        this.algorithm = zlib[this.algorithm];
+        if (!this.algorithm) throw new Error('Algorithm not found in zlib');
```
Comment (bebraw) [inline src/index.js]: This is another case I would rather force by linting. `if (...) {}` always as it bites too easily.

Comment (michael-ciniawsky) [inline src/index.js]: Newlines between logical/lexical blocks is 👍 imho. Better more then none in general 😛

File: package.json [hunk 3]
```diff
@@ -21,5 +53,19 @@
...
   },
-  "license": "MIT"
+  "license": "MIT",
+  "engines": {
+    "node": ">= 4.3 < 5.0.0 || >= 5.10"
+  },
+  "main": "dist/cjs.js",
+  "files": [
+    "dist"
+  ],
```
Comment (michael-ciniawsky) [inline package.json]:
```
"repository": {
    "type": "git",
    "url": "git+https://github.com/webpack-contrib/compression-webpack-plugin.git"
  },
  "bugs": {
    "url": "https://github.com/webpack-contrib/compression-webpack-plugin/issues"
  },
  "homepage": "https://webpack.js.org",
```

File: src/index.js (hunk 2 above)
Comment (d3viant0ne) [inline src/index.js]: @bebraw - last i check we do enforce `if (...) {}` with the latest version of the eslint preset.

Comment (bebraw) [inline src/index.js]: @d3viant0ne Ok, great. 👍

**Comment by d3viant0ne** [comment]
Need to fix a few things in the pacakge.json / readme before this merges.

**Comment by michael-ciniawsky** [comment]
`nodemon` && `serve:dev` 🔪  since `webpack-defaults` >= 0.4.6, I can take the README aswell but I'm not figthing for it 😛

**Comment by d3viant0ne** [comment]
It's a local branch, if you want to throw your repo standards changes in here go nuts.

**Pull request merged**